### PR TITLE
[FLOC-1570] Allow expressing ignorance about manifestations and applications on a Node

### PR DESCRIPTION
**Pull request by itamarst**
https://clusterhq.atlassian.net/browse/FLOC-1570

### CODE REVIEW

File: flocker/control/_model.py [hunk 1]
```diff
@@ -301,9 +347,8 @@ def __invariant__(self):
         return (True, "")
 
     hostname = field(type=unicode, factory=unicode, mandatory=True)
-    applications = pset_field(Application)
-    manifestations = field(type=PMap, initial=pmap(), factory=pmap,
-                           mandatory=True)
+    applications = pset_field(Application, optional=True)
+    manifestations = pmap_field(unicode, Manifestation, optional=True)
```
Comment (exarkun) [inline flocker/control/_model.py]: I don't see how the `__invariant__` can succeed if someone actually tries to instantiate a `Node` without one of these fields.

File: flocker/control/_model.py [hunk 2]
```diff
@@ -17,27 +17,71 @@
...
+    :param value: The required type for the values of the map.
+    :param bool optional: If true, ``None`` can be used as a value for
+        this field.
+
+    :return: A ``field`` containing a ``CheckedPMap``.
+    """
+    class TheMap(CheckedPMap):
+        __key_type__ = key_type
+        __value_type__ = value_type
+    TheMap.__name__ = key_type.__name__ + value_type.__name__ + "PMap"
```
Comment (exarkun) [inline flocker/control/_model.py]: I'm tempted to say something about underscores.  What's worse, `unicodeFilePathPMap` or `unicode_FilePath_PMap`?  Hard to say.  :question:

File: flocker/control/_model.py (hunk 1 above)
Comment (exarkun) [inline flocker/control/_model.py]: Probably means a few more unit tests would be appropriate to cover what we want to happen here, and then corresponding `__invariant__` fixes if they prove necessary.

Comment (exarkun) [inline flocker/control/_model.py]: Do we actually want these changes to be applied to `Node` or do they actually belong on `NodeState`?  The goal is to let the dataset convergence agent send back just manifestations and the container convergence agent send back just applications, right?  `P2PNodeDeployer.discover_local_state` returns a `NodeState`, not a `Node`.

**Comment by exarkun** [comment]
Thanks.  One of us is confused about `Node` vs `NodeState` here, I think.  Please explain or fix and re-submit.

File: flocker/control/_model.py (hunk 1 above)
Comment (itamarst) [inline flocker/control/_model.py]: We will definitely want them on `NodeState`, but I was going to do that in followup branch.

It needs to be in `Node` because that's how current cluster state is communicated to agents, and the goal is to allow the container agent to say "oh, manifestation state is missing, I should wait until that's available".

Comment (exarkun) [inline flocker/control/_model.py]: Ah, indeed.  Thanks, that makes sense.

Comment (itamarst) [inline flocker/control/_model.py]: Oh, and `optional=True` means "None is allowed". It doesn't mean you can omit the attribute (that's pyrsistent terminology, I reused it for consistency). The attribute still needs to be set.

Comment (itamarst) [inline flocker/control/_model.py]: So there needs to be test and invariant needs to be fixed, but it needs to handle `None`, not absence of attribute.

Comment (exarkun) [inline flocker/control/_model.py]: > We will definitely want them on NodeState, but I was going to do that in followup branch.

Which issue will that be?  FLOC-1570 just says "model objects" which seems to suggest this branch should complete the work.  I don't mind if there's another one to do the rest but maybe update the jira issue to describe the actual intended scope of this work.

Comment (itamarst) [inline flocker/control/_model.py]: As part of reworked proposal for FLOC-1542.

**Comment by exarkun** [comment]
Okay.  The `__invariant__` makes sense if we apply it to configuration only.

Do you want to change ... `ClusterStateService.as_deployment` to not use `to_node` as part of this branch?  Is that all that's necessary to switch to using `NodeState` instead of `Node` for representing cluster state in the control -> agent communication?  (I'm sure it couldn't be that easy...)

**Comment by itamarst** [comment]
The invariant applies to state as well... but I still think there's no way state could or should express that an application has a manifestation if it doesn't have knowledge of manifestations on the node.

`as_node()` is insufficient insofar as `NodeState.manifestations` is not a mapping. But that's probably easy to fix so I'll try to do that.

**Comment by exarkun** [comment]
> The invariant applies to state as well... but I still think there's no way state could or should express that an application has a manifestation if it doesn't have knowledge of manifestations on the node.

Ideally, yes.  And I agree this probably isn't something we **should** express (or want to).  I don't think I agree it's something we **couldn't** express - in particular, unintentionally.  We're trying to change `NodeState` so that it is constructed by two separate pieces of a distributed system.  I think if we want the invariant that `Node.__invariant__` implements to apply to state as well then we probably have more work to do.  Otherwise when some part of the system violates the invariant, the result is that pyrsistent refuses to let us represent the information and the system goes down some untested, unconsidered code path.

**Comment by itamarst** [comment]
If we express cluster state using a different class (`NodeState`) then it needn't have an invariant. So that solves that problem too?

**Comment by exarkun** [comment]
> If we express cluster state using a different class (NodeState) then it needn't have an invariant. So that solves that problem too?

Yes, I think that makes sense (maybe I misinterpreted your previous comment, I thought "The invariant applies to state as well" meant you wanted a similar `NodeState.__invariant__`).

**Comment by itamarst** [comment]
OK, will do.

File: flocker/control/_model.py [hunk 3]
```diff
@@ -284,13 +329,14 @@ class Node(PRecord):
...
-    :ivar frozenset applications: A ``frozenset`` of ``Application`` instances
-        describing the applications which are to run on this ``Node``.
+    :ivar applications: A ``PSet`` of ``Application`` instances describing
+        the applications which are to run on this ``Node``.
 
     :ivar PMap manifestations: Mapping between dataset IDs and
         corresponding ``Manifestation`` instances that are present on the
         node. Includes both those attached as volumes to any applications,
-        and those that are unattached.
+        and those that are unattached.w ``None`` if this information is
```
Comment (wallrj) [inline flocker/control/_model.py]: Typo "unattached.w"

File: flocker/control/_model.py [hunk 4]
```diff
@@ -17,27 +17,72 @@
...
 from pyrsistent import (
     pmap, PRecord, field, PMap, CheckedPSet, CheckedPMap,
+    optional as optional_type
     )
 
 from zope.interface import Interface, implementer
 
 
-def pset_field(klass):
+def pset_field(item_type, optional=False):
```
Comment (exarkun) [inline flocker/control/_model.py]: `optional` is too close to "the opposite of `mandatory`" for this API to be very clear.  How about `allow_none` or something similar?

File: flocker/control/_model.py [hunk 5]
```diff
@@ -17,27 +17,72 @@
...
+        def factory(argument):
+            if argument is None:
+                return None
+            else:
+                return TheSet(argument)
+    else:
+        factory = TheSet
+    return field(type=optional_type(TheSet) if optional else TheSet,
+                 factory=factory, mandatory=True,
                  initial=TheSet())
```
Comment (exarkun) [inline flocker/control/_model.py]: Do `mandatory=True` and `initial=...` make sense when combined?

File: flocker/node/test/test_loop.py [hunk 6]
```diff
@@ -37,6 +39,18 @@ def build_protocol():
...
+def to_node(node_state):
+    """
+    Convert a ``NodeState`` to a corresponding ``Node``.
+
+    :param NodeState node_state: Object to convert.
+    :return Node: Equivalent node.
+    """
+    return Node(hostname=node_state.hostname,
+                applications=node_state.applications,
+                manifestations=node_state.manifestations)
```
Comment (wallrj) [inline flocker/node/test/test_loop.py]: What if NodeState.applications is None? Isn't Node going to blow up?

Comment (wallrj) [inline flocker/node/test/test_loop.py]: Same goes for manifestations.

**Comment by wallrj** [comment]
Thanks @itamarst 

Please:
 * merge forward
 * resolve trivial test failures in test_blockdevice
 * answer or address comments above.

Then merge when green.

File: flocker/control/_model.py (hunk 4 above)
Comment (itamarst) [inline flocker/control/_model.py]: `optional` is what pyrsistent calls it. The pyrsistent author has already stated he's willing to accept this upstream, so I decided to stick to what would make sense there.

File: flocker/control/_model.py (hunk 5 above)
Comment (itamarst) [inline flocker/control/_model.py]: Yeah, the combination means you can't *remove* the attribute.

File: flocker/node/test/test_loop.py (hunk 6 above)
Comment (itamarst) [inline flocker/node/test/test_loop.py]: This is not a public API, this is a utility to make testing easier. Or rather, make tests that used to rely on `NodeState.to_node` continue to work. "Don't do that" basically :grinning:

**Comment by exarkun** [comment]
Test failures are spurious / unrelated.  Merging (so I can review #1247).
